### ISSUE TITLE
Reliable detection of new world generation vs. loading existing worlds

### ISSUE THREAD

**Issue by kno10** [minetest/minetest#15588]
### Problem

Games might add some functionality (e.g., new biomes) in a new version.
For old versions, they might want to fall back to old settings for compatibility to avoid certain breakages.

This is best handled by storing a "version at world generation" with the map, or using mod_storage (untested what works best - does `core.set_mapgen_setting` allow custom keys?)

However, users may still have legacy worlds where this value was not set. In order to distinguish the case of new world creation, vs. first use of an old world with the new mod version, it would be good to have a clear indication of whether a new world is currently generated.

Now `minetest.get_gametime()` sounds good at first - if the gametime is close to zero, its likely a new world.
This has also been discussed here: <https://forum.luanti.org/viewtopic.php?t=27682>
> As I mentioned in the original post, get_gametime returned nil during the initialization phase
It seems that this value is available too late to decide on which biomes to enable.

Related to: #5332

### Solutions

Some API that indicates whether the world is new or not.
Could be as simple as `core.is_new_world()` that returns `true` only on the first launch.

### Alternatives

A built-in API that tracks mod version numbers automatically.

**Comment by kromka-chleba** [comment]
>However, users may still have legacy worlds where this value was not set. In order to distinguish the case of new world creation, vs. first use of an old world with the new mod version, it would be good to have a clear indication of whether a new world is currently generated.

If your game has no backward/forward compatibility mechanism in the first place then your proposed solution won't really help here. Say you have 2 worlds, one generated with v0.0.1 of your game, another with v0.5.0. Between the releases you introduced 5 biomes. Then in v0.6.0 you added a version string in mod storage. Both the v0.0.1 and v0.5.0 worlds would be considered not new but otherwise you have no other info. What's worse is that the solution gives a false sense of knowing anything about the world.

Perhaps you could use `core.get_gametime()` to tell you if the world is new or old. A new world will have this value lower than 10 seconds. But you still get no version info.

> Could be as simple as core.is_new_world() that returns true only on the first launch.

What if the server never shuts down? The map would be eternally new.

**Comment by kno10** [comment]
> If your game has no backward/forward compatibility mechanism in the first place then your proposed solution won't really help here. Say you have 2 worlds, one generated with v0.0.1 of your game, another with v0.5.0. Between the releases you introduced 5 biomes. Then in v0.6.0 you added a version string in mod storage. Both the v0.0.1 and v0.5.0 worlds would be considered not new but otherwise you have no other info. What's worse is that the solution gives a false sense of knowing anything about the world.

Yes, but if a mod wants to introduce such a mechanism, this would still be better than nothing.
Some mod may not have such breaking changes for a few version, but plans on doing so.
It is *not* sufficient to just introduce the version info in one version and have everybody upgrade in two steps, because the changes need to be tracked per world. So you would have to require every player to open every world with some version 0.5.0 to adds a "0.5.0 or earlier" version info, only then upgrade to 0.6.0 and never touch an old world with it.

> Perhaps you could use `core.get_gametime()` to tell you if the world is new or old. A new world will have this value lower than 10 seconds. But you still get no version info.

According to the linked forum post above, the function returns `nil` during initialization, where you set up biomes!

> > Could be as simple as core.is_new_world() that returns true only on the first launch.
> 
> What if the server never shuts down? The map would be eternally new.

It will not load a new version of a mod, so for the purpose above this is fine. Call it `core.is_first_launch_of_world()` if you want.

**Comment by kno10** [comment]
I have verified, `core.get_gametime()` returns `nil` during mod initialization.
So it cannot be used to detect a new world during setup.

**Comment by kno10** [comment]
We could possibly hack and try to read the gametime from `env_meta.txt`...
it's horrible, but if Luanti does not add a better way to detect a new world, this is the best we can do.

**Comment by kromka-chleba** [comment]
>Perhaps you could use core.get_gametime() to tell you if the world is new or old. A new world will have this value lower than 10 seconds. But you still get no version info.

Btw better than this would be getting date of world creation which at least would give you some information about the possible version used by mapgen back then. Something like `core.get_world_creation_date()`.

Though I still fail to understand how would you reliably use that information? Use an old version of mapgen if the world is considered "old"? Or tell the player their world could get "destroyed"? Players usually want to get the new stuff on their old worlds. With my proposed `core.get_world_creation_date()` you'd still only be able to tell that the world uses mapgen version at that particular date or way older and nothing precise.

In the end mapgen versions can be stored in mod storage and an old world can be distinguished from a new world by using mod storage. So if you really care about these two you can easily implement them and use them. Otherwise I don't see why would devs spend their time writing a backward/forward compatibility solution for games that don't care about compatibility in the first place.

**Comment by kno10** [comment]
Calendar date is poor as you do not know when people update!
Plus: this again only works for new worlds, not for "legacy" worlds of your players!

Use case is simple: you are preparing some major changes. Examples could include:
- lowering the "bedrock" level
- introducing new biomes
- using certain new features (weighted biomes) that could cause ugly map inconsistencies
But your old version did not store version information.
Players *can* easily bump the mapgen version in their settings files to get the new stuff. These are text files.

Other things - such as new structures - may not need versioning. These can generate without causing ugly map inconsistencies.

Note that **you cannot force users to upgrade to an interim version first and load every single world to insert a version number prior to upgrading to the newest version**!

So either you introduce a version numbering scheme, wait for a year or longer, and whoever did not touch every of his worlds in the meantime has poor luck. Or you use the hack I describe above until the information is available via the Luanti API more elegantly...
